tchannel/tornado/connection.py

```diff
@@ -701,3 +633,162 @@ def _request_timed_out(self, req_id, req_ttl, future):
         self._request_tombstones.add(req_id, req_ttl)
+
+
+class Reader(object):
+
+    def __init__(self, io_stream, capacity=64):
+        self.queue = tornado.queues.Queue(capacity)
+        self.filling = False
+        self.io_stream = io_stream
+
+    def fill(self):
+        self.filling = True
+
+        io_loop = IOLoop.current()
+
+        def keep_reading(f):
+            if f.exception():
+                return log(f.exception())
+            # connect these two in the case when put blocks
+            self.queue.put(f.result())
+            io_loop.spawn_callback(self.fill)
+
+        io_loop.add_future(self._dequeue(), keep_reading)
+
+    def get(self):
+        """Receive the next message off the wire.
+
+        :returns:
+            A Future that produces a Context object containing the next
+            message off the wire.
+        """
+        if self.filling is False:
+            self.fill()
+
+        return self.queue.get()
+
+    def _dequeue(self):
+        # This is the message_future we'll return for any inbound messages.
+        message_future = tornado.gen.Future()
+        io_loop = IOLoop.current()
+
+        def on_body(read_body_future, size):
+            if read_body_future.exception():
+                return on_error(read_body_future)
+
+            body = read_body_future.result()
+            f = frame.frame_rw.read(BytesIO(body), size=size)
+            message_rw = messages.RW[f.header.message_type]
+            message = message_rw.read(BytesIO(f.payload))
+            message.id = f.header.message_id
+            message_future.set_result(message)
+
+        def on_read_size(read_size_future):
+            if read_size_future.exception():
+                return on_error(read_size_future)
+
+            size_bytes = read_size_future.result()
+            size = frame.frame_rw.size_rw.read(BytesIO(size_bytes))
+            read_body_future = self.io_stream.read_bytes(size - size_width)
+
+            io_loop.add_future(
+                read_body_future,
+                lambda future: on_body(future, size)
+            )
+
+            return read_body_future
+
+        def on_error(future):
+            log.info("Failed to read data: %s", future.exception())
+
+        size_width = frame.frame_rw.size_rw.width()
+        read_bytes_future = self.io_stream.read_bytes(size_width)
+        io_loop.add_future(read_bytes_future, on_read_size)
+
+        return message_future
+
+
+class Writer(object):
+
+    def __init__(self, io_stream, capacity=64):
+        self.queue = tornado.queues.Queue(capacity)
+        self.draining = False
+        self.io_stream = io_stream
+        # Tracks message IDs for this connection.
+        self._id_sequence = 0
+
+    def drain(self):
+        self.draining = True
+
+        io_loop = IOLoop.current()
+
+        def on_write(f, done):
+            if f.exception():
+                log.error(f.exception())
+                done.set_exception(f.exception())
```

Comment:
Prefer `set_exc_info` over `set_exception` to preserve stack traces in Python 2.